### ISSUE TITLE
Bring Back Option to Log to File

### ISSUE THREAD

**Issue by jdugan6240** [kakoune-lsp/kakoune-lsp#782]
Currently, kakoune-lsp always logs to the debug buffer. This can obscure debug messages from other Kakoune scripts, making it harder to debug those.

To make the debug buffer less noisy, kakoune-lsp should bring back the option to log to a file.

**Comment by krobelus** [comment]
Yeah that was a bit reckless.
Of course the idea was that, by default, only errors and some rare events should be printed.
This is the general convention for Kakoune plugins though I can see that it can be harder to fix errors if they come from a language server.
Currently we also print rare events like restarts but I'm not too attached to these.
Completely silencing all messages should be the last resort because otherwise people might notice only very late if `kak-lsp` or `rust-analyzer` are crashing again.

It should be possible to adjust log levels and silence individual language servers.

One issue is that we log all of the language server's stderr by default (at ERROR level).
I just noticed that `clangd` is somewhat chatty, it adds two lines for every request.
I didn't notice before because I usually redirect `clangd` stderr to a file.
So we might want to use `clangd --log=error`, perhaps dependent on the global kakoune-lsp log level.
Maybe that's too complicated and we should instead log the stderr at INFO level.
We can still detect crashing language servers (at least if they exit with SIGABRT) and print a single error line I guess. The stack trace should be available regardless, via the coredump.
So we'll probably add options and change the default log level from info to warn

	set-option lsp_log_file ""     # other options: /path/to/log, /dev/null (again, that's discouraged)
 	set-option lsp_log_level error # other options: warn, info, debug

Not yet sure about the log level change but I'd rather do that than too many people resort to setting `lsp_log_file`.
Long-term we should try to reduce reasons to ever use `lsp_log_file`.
Can you tell me which (error or non-error) logs you're seeing so we can fix them for everyone.
I have a couple of error logs mind that I haven't gotten around to yet.
Like rust-analyzer complaining about duplicate `textDocument/didOpen`, that's very easy to fix.

**Comment by krobelus** [comment]
ok in the `logs` branch I silenced some logs and added the `lsp_log_file` option.

Unfortunately there are a few language servers that are still somewhat noisy.
We may want to create a separate scratch `*lsp-log*` buffer for things that depend directly on the language server.
Other messages like panics should still go into `*debug*` (ideally even with the log file option but I'm not sure about the UI for that).

I guess for now the log file will do.

**Comment by jdugan6240** [comment]
I like this idea, as the `*debug*` buffer is a lot cleaner on this branch without even needing to set `lsp_log_file`. Before, every TOML-encoded request and every message to and from the server was getting logged there, which ended up being quite noisy.

I personally would argue, though, that all the lsp logs should go in the `*lsp-log*` buffer you mentioned instead of `*debug*`, so as not to pollute the `*debug*` buffer should the user choose a more permissive log level. IMO the `*debug*` buffer is meant to be used by Kakoune itself to log warnings and errors when running Kakoune scripts, as well as for the `echo -debug` and `debug` commands, to help debug said scripts. By writing kakoune-lsp logs to that same buffer, we're unnecessarily mixing debug messages from two different processes.

The problem with this, though, is the `*debug*` buffer is the only one that supports echo-ing directly to the buffer. I suppose we could use a FIFO for this (yet another FIFO, I know, but I don't see another way).

**Comment by krobelus** [comment]
> Before, every TOML-encoded request and every message to and from the server was getting logged there, which ended up being quite noisy.

This sounds a lot like a user error.
The `*debug*` buffer never got all requests/responses unless you set `lsp_debug=true` or passed `-vvv`.
That kind of verbosity was never meant for running in production beyond debugging something because it will fill up your disk (or at least RAM if on tmpfs) fairly quickly.

> I personally would argue, though, that all the lsp logs should go in the
> *lsp-log* buffer you mentioned instead of *debug*, so as not to pollute
> the *debug* buffer should the user choose a more permissive log level. IMO
> the *debug* buffer is meant to be used by Kakoune itself to log warnings and
> errors when running Kakoune scripts, as well as for the echo -debug and debug
> commands, to help debug said scripts. By writing kakoune-lsp logs to that same
> buffer, we're unnecessarily mixing debug messages from two different processes.

There have been lots of bug reports whose resolution could have been greatly expedited if the user had just checked the `*debug*` buffer.
It is the canonical place to find errors. Both errors from Kakoune core commands and plugins (whether the plugins live in the Kakoune repo or not).

I don't think kakoune-lsp should be fundamentally different from other plugins.
If I could I'd get rid of the compile step and make it a script.
The fact that it's a long-running daemon is also something I'd change if I could but it's kind of imposed by LSP.
So both are implementation details.

Another thing that perhaps biases me towards the `*debug*` buffer is that I have a mapping to quickly switch to it. Simple and effective.

I agree that in future, verbose logs should probably go to a dedicated LSP
buffer (similar to what Neovim or Helix do). But we should still export the errors to the `*debug*` buffer both for higher visibility and teaching users to use that.

> I suppose we could use a FIFO for this (yet another FIFO, I know, but I don't see another way).

yeah a fifo buffer would work well, it also means that we don't need `kak -p` processes for communicating logs

**Comment by jdugan6240** [comment]
> This sounds a lot like a user error. The *debug* buffer never got all requests/responses unless you set lsp_debug=true or passed -vvv.

Yep, my bad. I had -vvv set for debugging purposes a long time ago and just forgot to change it.

I do agree that errors should go in the `*debug*` buffer, since that's where other plugins' error data goes, and that more verbose logging could go to a different buffer. That seems like a good compromise.

**Comment by sclu1034** [comment]
I think that the specific bit "to file" in the initial post still holds merit.
All the extra stuff that would go to a separate buffer instead of `*debug*` is likely going come from `-vvv`, i.e. related to debugging a particular scenario. For those, it will likely be very useful to keep the information even after stopping a Kakoune session, and probably even useful to retain data from multiple sessions within a single file.

It doesn't have to be the default, but I think having an easy way to point the output to a file that gets opened in write-append mode would be very useful.

**Comment by krobelus** [comment]
I have addressed the immediate issue by ignoring the `-v` option.

I think most people will agree that we eventually want
1. an obvious way of collecting logs in a central place.
2. a way to view logs of all LSP requests/responses in a way that doesn't interfere with other workflows.
   Other editors record this log upfront.
   So as soon as you notice that LSP is broken again, you can switch to the LSP-log buffer and get at least the conversation history.
   Until we have that, our current approach is "enable debug logging and reproduce the issue".

Finally we do not currently have log categories (like Kakoune's `debug` option) or log levels anymore.
We might want to add those.

The current approach is very primitive but it works okay.

> I think having an easy way to point the output to a file that gets opened in write-append mode would be very useful.

Note that the `--log` file option still works. Logs are printed to both that file and the `*debug*` buffer, because I didn't
want to rob users who set `--log` (which used to be best practice) of the opportunity to discover that the `*debug*` buffer is all they need in most cases.

Of course the `--log` option is hidden so this is something to be improved later.

I was gonna commit the log file option already but I'm think it needs a little bit of design that I haven't done yet.
I guess the log file should be a superset of the both the `*lsp-log*` buffer as well as the `*debug*` entries (minus the errors that happen on the Kakoune script side I guess).

I'll leave the `logs` branch around; if anyone uses it, I'd be very interested to know how/why.
